### PR TITLE
[alert_handler/top] Top-level integration of LPG feature

### PR DESCRIPTION
**Pull request by msfschaffner**
This integrates the alert handler LPG concept into the top-level.
The mapping of resets / clocks to LPGs to alerts works as follows:
1) topgen determines the number of unique LPGs in the system (unique in this context means a unique combination of clock group, reset name and reset domain).
2) the clock gated / reset asserted indication signals from clkmgr and rstmgr are wired to the LPGs at the top-level (this also incurs a conversion from structs to arrays since the alert handler works on array signals).
3) the clock gated / reset asserted indication signals are then combined and synchronized to the alert handler clock within the alert handler
4) topgen provides a mapping from alert to LPG that is used inside the alert handler to index the correct combined indication signal for each alert

Note that this is dependent on several block-level changes (#8296, #8418, #8419) and needs a big rebase (some CI errors are due to pending DV alignments). Only the last two commits are actually relevant for this PR.

### CODE REVIEW

File: util/topgen/merge.py
Comment (tjaychen) [inline util/topgen/merge.py:766]: this part might need to change slightly due to #8410, since each clock can now have its own clock group (to support the escalator clocks)

File: util/topgen/merge.py
Comment (tjaychen) [inline util/topgen/merge.py:759]: hmm... does this work? because this gets you what the primary clock is in the hjson but not the top level connection right..?

File: util/topgen/merge.py
Comment (tjaychen) [inline util/topgen/merge.py:778]: ah wait, i think this answers my earlier question.

File: util/topgen/merge.py
Comment (tjaychen) [inline util/topgen/merge.py:785]: i wonder if it makes sense to extend the current clocks class with this information.. or maybe it might make more sense to create an alerts class that pulls in clock objects. Then you can pick out specific clock signals without needing to reconstruct these.

File: hw/top_earlgrey/rtl/autogen/top_earlgrey.sv
Comment (tjaychen) [inline hw/top_earlgrey/rtl/autogen/top_earlgrey.sv:806]: if it's easy to do, should we take out the `clk` and `rst` prefixes here? (might not be easy..) because this can easily be mistaken for a long list of clock / reset assignments and that could cause some confusion.

**Review by tjaychen — COMMENTED** [review]
this looks pretty good to me. I have some questions regarding a few things, but generally this makes sense.

File: hw/top_earlgrey/rtl/autogen/top_earlgrey.sv
Comment (msfschaffner) [reply hw/top_earlgrey/rtl/autogen/top_earlgrey.sv:806]: indeed, these labels might be a bit confusing. let me see what I can do (I reused the clock/reset names since they were readily available in existing data structures.).

File: util/topgen/merge.py
Comment (msfschaffner) [reply util/topgen/merge.py:778]: yes, this is the place where we get the actual clock connection. the former statement that gets `block_clock` only gets the clock label from the IP Hjson.

File: util/topgen/merge.py
Comment (msfschaffner) [reply util/topgen/merge.py:759]: see next comment.

File: util/topgen/merge.py
Comment (msfschaffner) [reply util/topgen/merge.py:766]: right, we will have to adjust this accordingly.

File: util/topgen/merge.py
Comment (msfschaffner) [reply util/topgen/merge.py:785]: ok let me see... I believe it does not make too much sense to add it to the clock / reset classes, since the LPGs are formed when combining specific resets with a clock group. so probably it makes more sense to create an alert class to hold this info.

this will probably need some more refactoring of the alerts into class objects (besides the LPG stuff here).
I'll give this a try to see how much would have to change.

Comment (tjaychen) [reply util/topgen/merge.py:785]: o yeah i think this makes more sense as an alert class that references clocks / resets objects also.
i don't think we should bother now though..that feels likes a subsequent PR cleanup if we have time.

File: hw/top_earlgrey/rtl/autogen/top_earlgrey.sv
Comment (msfschaffner) [reply hw/top_earlgrey/rtl/autogen/top_earlgrey.sv]: I have changed this and removed the `clk_` and `rst_` labels.

File: util/topgen/merge.py
Comment (msfschaffner) [reply util/topgen/merge.py:785]: yeah let's get these connections in first.

**Comment by msfschaffner** [comment]
Alright, I have rebased this onto recent changes - PTAL.

**Comment by msfschaffner** [comment]
Rebased and updated such that the LPG logic takes into account per-clock clock group assignments (if they are present).

**Review by cindychip — COMMENTED** [review]
It LGTM, but I will defer to designers to approve it :)

**Comment by msfschaffner** [comment]
Rebased onto recent changes on master.

**Comment by tjaychen** [comment]
sorry for being late on reviews, i'll have a one last closer look and let's
get it merged.

**Comment by msfschaffner** [comment]
No worries!

Last push just rebased this onto the latest rstmgr changes.

File: hw/ip/alert_handler/data/alert_handler.hjson
Comment (tjaychen) [inline hw/ip/alert_handler/data/alert_handler.hjson:66]: nit, spacing seems a little weird here.

File: hw/ip/alert_handler/rtl/alert_handler_lpg_ctrl.sv
Comment (tjaychen) [inline hw/ip/alert_handler/rtl/alert_handler_lpg_ctrl.sv:33]: should we swap the order of these two? I think here we're basically putting combo logic ahead of the sync, and this might result in us getting one cycle glitches.  It's not a huge deal i think since this is multi-bit, and this situation could happen anyways, but it's a bit cleaner to make the sync paths flop to flop

Comment (msfschaffner) [reply hw/ip/alert_handler/rtl/alert_handler_lpg_ctrl.sv:33]: yeah as you see I had it that way before. 
I mainly did this in order to save flops since each synchronizer adds 8 more...
but since we're consolidating the LPGs now, this should not be such a big problem anymore.

File: hw/ip/alert_handler/rtl/alert_handler_lpg_ctrl.sv
Comment (tjaychen) [inline hw/ip/alert_handler/rtl/alert_handler_lpg_ctrl.sv:45]: do you think the default for this should be `ON`? maybe it doesn't make a difference.. i'm not sure at the moment. It feels more consistent with the default state of the `alert_receiver` fsm.

File: hw/ip/alert_handler/rtl/alert_handler_lpg_ctrl.sv
Comment (msfschaffner) [reply hw/ip/alert_handler/rtl/alert_handler_lpg_ctrl.sv:33]: let me revert this.

File: hw/ip/alert_handler/rtl/alert_handler_lpg_ctrl.sv
Comment (tjaychen) [inline hw/ip/alert_handler/rtl/alert_handler_lpg_ctrl.sv:49]: i think this is true, but it's worth walking through it one more time.
Since these are multi-bit syncs (with possible non-sense values in the middle), i think our argument is as follows..

When going from OFF -> ON, all the possible values except for "ON" are considered OFF, so we should not have a case of glitching between OFF->ON->OFF->ON.

When going from ON->OFF, all the possible values except for "ON" are OFF, so we should not have a case of glitching between ON->OFF->ON->OFF.

is that right?

File: hw/top_earlgrey/ip/alert_handler/data/autogen/alert_handler.hjson
Comment (tjaychen) [inline hw/top_earlgrey/ip/alert_handler/data/autogen/alert_handler.hjson:74]: same spacing nit

**Review by tjaychen — APPROVED** [review]
lgtm with a few clarifying questions and comments.

File: hw/ip/alert_handler/rtl/alert_handler_lpg_ctrl.sv
Comment (msfschaffner) [reply hw/ip/alert_handler/rtl/alert_handler_lpg_ctrl.sv:45]: yes, we can make this default to on (I've done that as well on the `rstmgr` / `clkmgr` side)

File: hw/ip/alert_handler/data/alert_handler.hjson
Comment (msfschaffner) [reply hw/ip/alert_handler/data/alert_handler.hjson:66]: I've aligned the template to print this out more nicely

File: hw/top_earlgrey/ip/alert_handler/data/autogen/alert_handler.hjson
Comment (msfschaffner) [reply hw/top_earlgrey/ip/alert_handler/data/autogen/alert_handler.hjson]: jep, this was a templating issue and is fixed now.

File: hw/ip/alert_handler/rtl/alert_handler_lpg_ctrl.sv
Comment (msfschaffner) [reply hw/ip/alert_handler/rtl/alert_handler_lpg_ctrl.sv]: reverted to use separate syncs.

File: hw/ip/alert_handler/rtl/alert_handler_lpg_ctrl.sv
Comment (msfschaffner) [reply hw/ip/alert_handler/rtl/alert_handler_lpg_ctrl.sv]: I've added a parameter to `prim_lc_sync` to set the reset value to `ON`

**Comment by msfschaffner** [comment]
Thanks for the reviews! Merging this now...

**Pull request merged**
